### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by KirilKabakchiev**

### CODE REVIEW

File: pkg/osb/broker_fetcher.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+// BrokerDetails implements osb.BrokerRoundTripper
+type BrokerDetails struct {
+	Username string
+	Password string
+	URL      string
+}
+
+var _ osb.BrokerFetcher = &BrokerDetails{}
```
Comment (NickyMateev) [inline pkg/osb/broker_fetcher.go]: Maybe we should move the BrokerFetcher interface to **pkg** folder in the SM project since we use the interface in other projects (like here)?

File: pkg/logging/error_location_hook_test.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+				actualErrorValue:         nil,
+				expectedFireError:        nil,
+				expectedErrorSourceValue: "",
+			}),
+
+			Entry("When error key field is not an error it returns an error", testCase{
+				actualErrorValue: func(err error, stacks, wraps int) interface{} {
+					return struct{ A string }{A: "val"}
+				},
+				expectedFireError:        fmt.Errorf("object logged as error does not satisfy error interface"),
```
Comment (dpanayotov) [inline pkg/logging/error_location_hook_test.go:88]: extract error

File: pkg/osb/broker_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+package osb
+
+import (
+	"context"
+	"github.com/Peripli/service-manager/api/osb"
+	"github.com/Peripli/service-manager/pkg/types"
+)
+
+// BrokerDetails implements osb.BrokerRoundTripper
+type BrokerDetails struct {
```
Comment (dpanayotov) [inline pkg/osb/broker_fetcher.go]: BrokerDetailsFetcher?

File: pkg/osb/broker_fetcher.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+package osb
+
+import (
+	"context"
+	"github.com/Peripli/service-manager/api/osb"
+	"github.com/Peripli/service-manager/pkg/types"
+)
+
+// BrokerDetails implements osb.BrokerRoundTripper
```
Comment (dpanayotov) [inline pkg/osb/broker_fetcher.go]: implements osb.BrokerFetcher

File: pkg/osb/broker_fetcher.go (hunk 1 above)
Comment (dpanayotov) [inline pkg/osb/broker_fetcher.go]: do we still need to do this? I mean it's redundant - if it's not gonna compile here, it's not gonna compile in the place that depends on this

File: pkg/sbproxy/options.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"github.com/Peripli/service-broker-proxy/pkg/sbproxy/reconcile"
+	"github.com/Peripli/service-broker-proxy/pkg/sm"
+	"github.com/Peripli/service-manager/pkg/env"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/server"
+	"github.com/spf13/pflag"
+)
+
+type validatable interface {
```
Comment (dpanayotov) [inline pkg/sbproxy/options.go]: service-manager/pkg/util/api.go InputValidator

File: pkg/sbproxy/options.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	Validate() error
+}
+
+// Settings type holds all config properties for the sbproxy
+type Settings struct {
+	Server    *server.Settings    `mapstructure:"server"`
+	Log       *log.Settings       `mapstructure:"log"`
+	Sm        *sm.Settings        `mapstructure:"sm"`
+	Reconcile *reconcile.Settings `mapstructure:"self"`
```
Comment (dpanayotov) [inline pkg/sbproxy/options.go]: Can this be package app and then the settings will be app.Settings instead of self.Settings? app.NewReconcilationTask still works IMO

File: pkg/sbproxy/sbproxy_test.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+						set.Set("sm.user", "")
+						set.Set("sm.password", "admin")
+						set.Set("sm.url", "http://localhost:8080")
+						set.Set("sm.osb_api_path", "/osb")
+					}), fakePlatformClient)
+				}).To(Panic())
+			})
+		})
+
+		//Context("when adding cron job fails due to invalid schedule", func() {
```
Comment (dpanayotov) [inline pkg/sbproxy/sbproxy_test.go]: is this to be fixed or to be removed?

File: pkg/sbproxy/reconcile/reconcile_brokers_test.go [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+		fakePlatformCatalogFetcher.FetchReturns(nil)
+	}
+
+	BeforeEach(func() {
+		fakeSMClient = &smfakes.FakeClient{}
+		fakePlatformBrokerClient = &platformfakes.FakeClient{}
+		fakePlatformCatalogFetcher = &platformfakes.FakeCatalogFetcher{}
+		fakePlatformServiceAccess = &platformfakes.FakeServiceAccess{}
+
+		fakeWG = &sync.WaitGroup{}
```
Comment (dpanayotov) [inline pkg/sbproxy/reconcile/reconcile_brokers_test.go]: it's real, not fake

File: pkg/sbproxy/reconcile/reconcile_brokers_test.go [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+		}
+
+		platformbrokerNonProxy = platform.ServiceBroker{
+			GUID:      "platformBrokerID3",
+			Name:      "platformBroker3",
+			BrokerURL: "https://platformBroker3.com",
+		}
+	})
+
+	type reconcilationExceptations struct {
```
Comment (dpanayotov) [inline pkg/sbproxy/reconcile/reconcile_brokers_test.go]: expectations

File: pkg/sbproxy/reconcile/reconcile_brokers_test.go [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+					reconcileCreate:  []platform.ServiceBroker{},
+					reconcileDelete:  []platform.ServiceBroker{},
+					reconcileCatalog: []platform.ServiceBroker{},
+					reconcileAccess:  []osbc.CatalogResponse{},
+				}
+			},
+		}),
+	}
+
+	DescribeTable("Run", func(t testCase) {
```
Comment (dpanayotov) [inline pkg/sbproxy/reconcile/reconcile_brokers_test.go:390]: Can this be above the table entries? It seems more natural to first know what is the test and then see the variations, because reading from top to bottom you read first the table entry without knowing what's the purpose of it

File: pkg/sbproxy/reconcile/reconcile_brokers_test.go [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+			},
+			smBrokers: func() ([]sm.Broker, error) {
+				return []sm.Broker{
+					smbroker1,
+				}, nil
+			},
+			expectations: func() reconcilationExceptations {
+				return reconcilationExceptations{
+					reconcileCreate: []platform.ServiceBroker{
+						platformbroker1,
```
Comment (dpanayotov) [inline pkg/sbproxy/reconcile/reconcile_brokers_test.go:243]: This is a bit hard to follow. These expectations are not what the result of the execution would be, but what are the parameters passed to the operation. Better call them something like "opArgsExpectations" with "createCalledWith"

File: pkg/sbproxy/reconcile/reconcile_brokers.go [hunk 12]
```diff
@@ -81,102 +113,100 @@ func (r ReconcileBrokersTask) run() {
 
 // reconcileBrokers attempts to reconcile the current brokers state in the platform (existingBrokers)
 // to match the desired broker state coming from the Service Manager (payloadBrokers).
-func (r ReconcileBrokersTask) reconcileBrokers(existingBrokers []serviceBrokerReg, payloadBrokers []serviceBrokerReg) {
+func (r ReconcilationTask) reconcileBrokers(existingBrokers []platform.ServiceBroker, payloadBrokers []platform.ServiceBroker) {
```
Comment (dpanayotov) [inline pkg/sbproxy/reconcile/reconcile_brokers.go:116]: There is a Broker and a ServiceBroker (one is from SM, one is from Platform) Can we unify them somehow?

File: pkg/sbproxy/reconcile/reconcile_brokers.go [hunk 13]
```diff
@@ -81,102 +113,100 @@ func (r ReconcileBrokersTask) run() {
...
+		delete(existingMap, payloadBroker.GUID)
 		if existingBroker == nil {
-			r.createBrokerRegistration(&payloadBroker.ServiceBroker)
+			r.createBrokerRegistration(&payloadBroker)
 		} else {
-			r.fetchBrokerCatalog(&existingBroker.ServiceBroker)
+			r.fetchBrokerCatalog(existingBroker)
 		}
-		r.enableServiceAccessVisibilities(&payloadBroker.ServiceBroker)
+		r.enableServiceAccessVisibilities(&payloadBroker)
```
Comment (dpanayotov) [inline pkg/sbproxy/reconcile/reconcile_brokers.go]: this can be skipped if any of the above operations failed

**Review by dpanayotov — CHANGES_REQUESTED** [review]

File: pkg/sbproxy/reconcile/reconcile_brokers_test.go (hunk 10 above)
Comment (KirilKabakchiev) [reply pkg/sbproxy/reconcile/reconcile_brokers_test.go:390]: it wont compile as we use entries as argument to describeTable

File: pkg/sbproxy/reconcile/reconcile_brokers_test.go (hunk 11 above)
Comment (KirilKabakchiev) [reply pkg/sbproxy/reconcile/reconcile_brokers_test.go:243]: Something like reconcileCreateCalledFor?

File: pkg/sbproxy/reconcile/reconcile_brokers.go (hunk 12 above)
Comment (KirilKabakchiev) [reply pkg/sbproxy/reconcile/reconcile_brokers.go:116]: They slightly differ (e.g. platform brokers have a name that we use as well as part of the delete request. I could embed sm broker into platform broker but this will get us cycle deplendencies. I will leave it like this for now

File: pkg/osb/broker_fetcher.go (hunk 1 above)
Comment (KirilKabakchiev) [reply pkg/osb/broker_fetcher.go]: It let's you see what implementations we care about, seems useful for readability

File: pkg/sbproxy/options.go (hunk 6 above)
Comment (KirilKabakchiev) [reply pkg/sbproxy/options.go]: it seems a little weird - in app package to have just a reconcilation task - i can switch the mapstructure to tag it as app though

**Review by dpanayotov — APPROVED** [review]
